### DISCUSSION TITLE
Rate limiting when behind a proxy or load balancer

### DISCUSSION THREAD

**Discussion by edclement** [verdaccio/verdaccio#3269] [Ideas]
@juanpicado - I'm interesting in getting your feedback on this discussion

### Issue
I did some research on how the rate limiting within Verdaccio is actually implemented (`5.x` branch). 

Rate limiting is implemented for the following endpoints:
1. [`/-/verdaccio/data/*`](https://github.com/verdaccio/verdaccio/blob/v5.13.3/src/api/web/endpoint/index.ts#L16) - This is feeding data to the web UI. Examples include the list of npm packages and data related to individual packages.
2. [`/-/verdaccio/sec/*`](https://github.com/verdaccio/verdaccio/blob/v5.13.3/src/api/web/endpoint/index.ts#L21) - This is used for the user login/logout from the web UI.
3. [`/user`](https://github.com/verdaccio/verdaccio/blob/v5.13.3/src/api/endpoint/api/v1/profile.ts#L36) - Registry user profile operations
4. [`/tokens/*`](https://github.com/verdaccio/verdaccio/blob/v5.13.3/src/api/endpoint/api/v1/token.ts#L30) - Registry token related operations (e.g. get/create/delete tokens).
5. [`/-/user/*`](https://github.com/verdaccio/verdaccio/blob/v5.13.3/src/api/endpoint/api/user.ts#L25) - Registry user login/logout operations.

The [`express-rate-limit`](https://www.npmjs.com/package/express-rate-limit) plugin is used to perform rate limiting (specifically version `5.5.1` in Verdaccio version `5.13.3`). This plugin, by default, relies on the `request.ip` field as the client/user key to perform rate limiting (see [here](https://github.com/nfriedly/express-rate-limit/blob/v5.5.1/lib/express-rate-limit.js#L20)). Express resolves the IP of the request by default as `req.socket.remoteAddress` **unless** the configuration option `trust proxy` is set and an IP (or set of IPs) to trust is passed (see express docs [here](https://expressjs.com/en/guide/behind-proxies.html)). In that case, the `X-Forwarded-For` header is relied on to derive the client/user IP.

The `trust proxy` option is not being set by Verdaccio (based on investigation/searching) and there is no Verdaccio configuration setting that allows this option to be set either. Setting this option would also involve passing the IP of the proxy/load-balancer that should be trusted.

What this means, is that given the current implementation of rate limiting in Verdaccio, the rate limiting will be applied to the IP of the reverse-proxy or load-balancer Verdaccio is placed behind and not the actual requesting IP contained in the `X-Forwarded-For` header. For this reason, it is dangerous to apply rate limiting configuration to Verdaccio when it is placed behind a reverse-proxy or load-balancer.

### Proposal
I propose opening a PR where configuration can be passed to Verdaccio that allows:
```
app.set('trust proxy', 'X.X.X.X');
```
...API call to be invoked.

I believe the configuration value for this in Verdaccio should be placed under the existing [server](https://verdaccio.org/docs/configuration/#server) section and look like:
```yml
server:
  trustProxy: loopback, 123.456.789.1
```

Some simple code could then get added to [`src/api/index.ts`](https://github.com/verdaccio/verdaccio/blob/v5.13.3/src/api/index.ts#L29):
```TypeScript
if (config.server?.trustProxy) {
  app.set('trust proxy', config.server.trustProxy);
}
```

Based on my knowledge of Verdaccio, I don't think there would be any negative consequences in other areas of the code of configuring this express setting. **I'm open to others thoughts on this matter though**.

Implementing this should actually improve logging when Verdaccio is placed behind a reverse-proxy or load-balancer since an accurate client IP will be used.

**Comment by juanpicado** [comment]
You are right, make sense.

**Reply by edclement** [reply]
If you are fine with my recommended approach, I will try to open a PR soon.

**Reply by juanpicado** [reply]
fyi, If you decide also do it on master (6.x) defer it for another moment, I'm rewriting https://github.com/verdaccio/verdaccio/pull/3100 everything now on master.

**Comment by edclement** [comment]
@juanpicado - I've started looking more closely at how to implement this in the `5.x` branch. Let me know if the approach I describe below makes sense or if you have a different recommendation.

It looks like the `config` object passed to the [`defineAPI` function in `src/api/index.ts`](https://github.com/verdaccio/verdaccio/blob/v5.13.3/src/api/index.ts#L23) is of type `IConfig`. That type is imported from from `@verdaccio/types`. Before I can make any code updates to the Verdaccio 5.x branch, a PR will need to be opened against the [verdaccio/monorepo]() project and the version for that will need to be bumped.

The [`ServerSettingsConfig`](https://github.com/verdaccio/monorepo/blob/%40verdaccio/types%4010.5.1/core/types/src/configuration.ts#L142) will need to be updated:
```TypeScript
export type ServerSettingsConf = {
  // docs on `trustProxy` can be found at: https://expressjs.com/en/guide/behind-proxies.html
  trustProxy?: string; 
  // express-rate-limit settings
  rateLimit: RateLimit;
  keepAliveTimeout?: number;
};
```

**Question**: Which branch on the monorepo project should those changes be based off of? Verdaccio 5.13.3 currently has a dependency on `@verdaccio/types@10.5.1`.

Also, I notice that the `ServerSettingsConfig` type has a `rateLimit` property defined. This configuration property (`server.rateLimit`) is not mentioned in the docs. Only [`userRateLimit`](https://verdaccio.org/docs/configuration/#user-rate-limit) and [`web.rateLimit`](https://verdaccio.org/docs/webui/#configuration) are defined in the docs and appear to be used in the Verdaccio code (from what I can tell). Is `server.rateLimit` a valid configuration property?

**Reply by juanpicado** [reply]
I have no computer now, but seems your are not the right track. Types might me outdated already price that by refactoring master branch. Answering question yes first in monorepo and then we can merge at 5.x

**Reply by juanpicado** [reply]
I have no computer now, but seems your are in the right track. Types might me outdated already price that by refactoring master branch. Answering question yes first in monorepo and then we can merge at 5.x

**Reply by edclement** [reply]
PR opened against the monorepo [here](https://github.com/verdaccio/monorepo/pull/604) targeting the `9.x` branch. I marked it as a patch version bump. Let me know if you prefer a minor version bump or if `9.x` is not the correct branch. Thanks!

**Reply by juanpicado** [reply]
The types dep was published 10.5.2

**Reply by edclement** [reply]
I opened a PR against Verdaccio: https://github.com/verdaccio/verdaccio/pull/3274. I verified manually (running Verdaccio) that setting the `trust proxy`  express option works as expected. However, I could not get the unit tests to respect this option. I'm guessing it has something to do with the use of [supertest](https://www.npmjs.com/package/supertest) for unit tests. I provided a more thorough explanation in the PR.

**Reply by juanpicado** [reply]
I think we can resolve this, right?
